File: kgk_customisations/kgk_customisations/report/audit_trail_report/audit_trail_report.js
// Copyright (c) 2025, KGK and contributors
// For license information, please see license.txt

frappe.query_reports["Audit Trail Report"] = {
	"filters": [
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_days(frappe.datetime.get_today(), -30),
			"reqd": 1
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1
		},
		{
			"fieldname": "activity_type",
			"label": __("Activity Type"),
			"fieldtype": "Select",
			"options": "\nAll\nBalance Update\nManual Verification\nERP Verification\nFinal Verification\nWorkflow Transition\nDocument Creation\nDocument Modification\nDocument Cancellation",
			"default": "All"
		},
		{
			"fieldname": "user",
			"label": __("User"),
			"fieldtype": "Link",
			"options": "User"
		},
		{
			"fieldname": "document_type",
			"label": __("Document Type"),
			"fieldtype": "Select",
			"options": "\nAll\nCash Document\nDaily Cash Balance\nCash Balance Submission\nBank Basic Entry"
		},
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company"
		},
		{
			"fieldname": "search_text",
			"label": __("Search in Details"),
			"fieldtype": "Data"
		}
	],
	
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		if (column.fieldname == "activity_type" && data && data.activity_type) {
			const type_colors = {
				"Balance Update": "blue",
				"Manual Verification": "green",
				"ERP Verification": "darkgreen",
				"Final Verification": "purple",
				"Workflow Transition": "orange",
				"Document Creation": "teal",
				"Document Modification": "gray",
				"Document Cancellation": "red"
			};
			const color = type_colors[data.activity_type] || "black";
			value = "<span style='color:" + color + "'>" + value + "</span>";
		}
		
		if (column.fieldname == "document_name" && data && data.document_name && data.document_type) {
			value = `<a href="/app/${data.document_type.toLowerCase().replace(/ /g, '-')}/${data.document_name}">${data.document_name}</a>`;
		}
		
		return value;
	},
	
	"onload": function(report) {
		report.page.add_inner_button(__("Export for Auditors"), function() {
			// Export with all details
			frappe.query_report.export_report('Excel');
		});
		
		report.page.add_inner_button(__("Activity Summary"), function() {
			show_activity_summary();
		});
		
		report.page.add_inner_button(__("User Activity Timeline"), function() {
			show_user_timeline();
		});
		
		report.page.add_inner_button(__("Compliance Check"), function() {
			run_compliance_check();
		});
		
		function show_activity_summary() {
			const data = frappe.query_report.data;
			if (!data || data.length === 0) {
				frappe.msgprint(__("No data available"));
				return;
			}
			
			// Group by activity type
			const summary = {};
			const user_summary = {};
			
			data.forEach(row => {
				const activity = row.activity_type || "Unknown";
				const user = row.user || "Unknown";
				
				summary[activity] = (summary[activity] || 0) + 1;
				
				if (!user_summary[user]) {
					user_summary[user] = {};
				}
				user_summary[user][activity] = (user_summary[user][activity] || 0) + 1;
			});
			
			let html = '<h4>Activity Summary</h4>';
			html += '<table class="table table-bordered"><thead><tr>';
			html += '<th>Activity Type</th><th>Count</th></tr></thead><tbody>';
			
			Object.keys(summary).sort((a, b) => summary[b] - summary[a]).forEach(activity => {
				html += `<tr><td>${activity}</td><td>${summary[activity]}</td></tr>`;
			});
			
			html += '</tbody></table>';
			
			html += '<h4 style="margin-top: 20px;">User Activity Summary</h4>';
			html += '<table class="table table-bordered"><thead><tr>';
			html += '<th>User</th><th>Total Actions</th><th>Details</th></tr></thead><tbody>';
			
			Object.keys(user_summary).forEach(user => {
				const total = Object.values(user_summary[user]).reduce((a, b) => a + b, 0);
				const details = Object.keys(user_summary[user])
					.map(act => `${act}: ${user_summary[user][act]}`)
					.join(', ');
				html += `<tr><td>${user}</td><td>${total}</td><td style="font-size: 0.9em;">${details}</td></tr>`;
			});
			
			html += '</tbody></table>';
			
			frappe.msgprint({
				title: __("Activity Summary"),
				message: html,
				wide: true
			});
		}
		
		function show_user_timeline() {
			const filters = frappe.query_report.get_filter_values();
			if (!filters.user) {
				frappe.msgprint(__("Please select a user first"));
				return;
			}
			
			const data = frappe.query_report.data;
			if (!data || data.length === 0) {
				frappe.msgprint(__("No data available"));
				return;
			}
			
			let html = '<div class="timeline">';
			
			data.forEach(row => {
				html += `<div class="timeline-item" style="margin-bottom: 15px; padding: 10px; border-left: 3px solid #007bff;">`;
				html += `<div style="font-weight: bold;">${row.timestamp}</div>`;
				html += `<div style="color: #007bff;">${row.activity_type}</div>`;
				html += `<div>${row.document_type}: ${row.document_name}</div>`;
				if (row.details) {
					html += `<div style="font-size: 0.9em; color: gray;">${row.details}</div>`;
				}
				html += `</div>`;
			});
			
			html += '</div>';
			
			frappe.msgprint({
				title: __("User Activity Timeline - ") + filters.user,
				message: html,
				wide: true
			});
		}
		
		function run_compliance_check() {
			const data = frappe.query_report.data;
			if (!data || data.length === 0) {
				frappe.msgprint(__("No data available"));
				return;
			}
			
			// Check for compliance issues
			const issues = [];
			
			// Check 1: Balance updates without verification
			const balance_updates = data.filter(r => r.activity_type === 'Balance Update');
			const verifications = data.filter(r => 
				r.activity_type === 'Manual Verification' || 
				r.activity_type === 'ERP Verification' || 
				r.activity_type === 'Final Verification'
			);
			
			if (balance_updates.length > verifications.length * 2) {
				issues.push(`Warning: ${balance_updates.length} balance updates but only ${verifications.length} verifications`);
			}
			
			// Check 2: Cancellations without proper authorization
			const cancellations = data.filter(r => r.activity_type === 'Document Cancellation');
			if (cancellations.length > 0) {
				issues.push(`Info: ${cancellations.length} document cancellations recorded`);
			}
			
			// Check 3: Same user creating and approving
			const workflow_transitions = data.filter(r => r.activity_type === 'Workflow Transition');
			// This would need more complex logic to check same-user approval
			
			let html = '<h4>Compliance Check Results</h4>';
			
			if (issues.length === 0) {
				html += '<p style="color: green;">✓ No compliance issues detected</p>';
			} else {
				html += '<ul>';
				issues.forEach(issue => {
					html += `<li>${issue}</li>`;
				});
				html += '</ul>';
			}
			
			html += '<p style="margin-top: 20px; font-style: italic; color: gray;">';
			html += 'Note: This is a basic compliance check. For detailed auditing, export the full report.';
			html += '</p>';
			
			frappe.msgprint({
				title: __("Compliance Check"),
				message: html,
				wide: true
			});
		}
	}
};

File: kgk_customisations/kgk_customisations/report/audit_trail_report/audit_trail_report.py
# Copyright (c) 2025, KGK and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate, get_datetime, now_datetime
from kgk_customisations.kgk_customisations.utils.query_builder import SafeQueryBuilder


def execute(filters=None):
	"""
	Audit Trail Report - Comprehensive compliance and activity tracking
	"""
	if not filters:
		filters = {}
	
	columns = get_columns()
	data = get_data(filters)
	
	return columns, data


def get_columns():
	"""Define audit trail report columns"""
	return [
		{
			"fieldname": "timestamp",
			"label": _("Timestamp"),
			"fieldtype": "Datetime",
			"width": 160
		},
		{
			"fieldname": "activity_type",
			"label": _("Activity Type"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "user",
			"label": _("User"),
			"fieldtype": "Link",
			"options": "User",
			"width": 150
		},
		{
			"fieldname": "document_type",
			"label": _("Document Type"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "document_name",
			"label": _("Document"),
			"fieldtype": "Dynamic Link",
			"options": "document_type",
			"width": 150
		},
		{
			"fieldname": "company",
			"label": _("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"width": 130
		},
		{
			"fieldname": "old_value",
			"label": _("Old Value"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "new_value",
			"label": _("New Value"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "details",
			"label": _("Details"),
			"fieldtype": "Text",
			"width": 250
		},
		{
			"fieldname": "ip_address",
			"label": _("IP Address"),
			"fieldtype": "Data",
			"width": 120
		}
	]


def get_data(filters):
	"""Fetch audit trail data from Cash Document Audit Trail"""
	query = """
		SELECT
			cat.timestamp,
			cat.activity_type,
			cat.user,
			cat.document_type,
			cat.document_name,
			cat.company,
			cat.old_value,
			cat.new_value,
			cat.details,
			cat.ip_address
		FROM `tabCash Document Audit Trail` cat
	"""
	
	# Build WHERE conditions safely
	conditions = []
	params = {}
	
	if filters.get("from_date"):
		conditions.append("DATE(cat.timestamp) >= %(from_date)s")
		params["from_date"] = filters.get("from_date")
	
	if filters.get("to_date"):
		conditions.append("DATE(cat.timestamp) <= %(to_date)s")
		params["to_date"] = filters.get("to_date")
	
	if filters.get("activity_type") and filters.get("activity_type") != "All":
		conditions.append("cat.activity_type = %(activity_type)s")
		params["activity_type"] = filters.get("activity_type")
	
	if filters.get("user"):
		conditions.append("cat.user = %(user)s")
		params["user"] = filters.get("user")
	
	if filters.get("document_type") and filters.get("document_type") != "All":
		conditions.append("cat.document_type = %(document_type)s")
		params["document_type"] = filters.get("document_type")
	
	if filters.get("company"):
		conditions.append("cat.company = %(company)s")
		params["company"] = filters.get("company")
	
	if conditions:
		query += " WHERE " + " AND ".join(conditions)
	
	query += " ORDER BY cat.timestamp DESC"
	
	data = frappe.db.sql(query, params, as_dict=1)
	
	# Apply text search filter if provided
	search_text = filters.get("search_text")
	if search_text:
		search_text = search_text.lower()
		data = [
			row for row in data
			if search_text in (row.get("details") or "").lower()
			or search_text in (row.get("old_value") or "").lower()
			or search_text in (row.get("new_value") or "").lower()
		]
	
	return data


def get_conditions(filters):
	"""Build SQL WHERE conditions"""
	conditions = []
	
	if filters.get("from_date"):
		conditions.append("DATE(cat.timestamp) >= %(from_date)s")
	
	if filters.get("to_date"):
		conditions.append("DATE(cat.timestamp) <= %(to_date)s")
	
	if filters.get("activity_type") and filters.get("activity_type") != "All":
		conditions.append("cat.activity_type = %(activity_type)s")
	
	if filters.get("user"):
		conditions.append("cat.user = %(user)s")
	
	if filters.get("document_type") and filters.get("document_type") != "All":
		conditions.append("cat.document_type = %(document_type)s")
	
	if filters.get("company"):
		conditions.append("cat.company = %(company)s")
	
	return " AND " + " AND ".join(conditions) if conditions else ""


@frappe.whitelist()
def export_audit_report(filters):
	"""
	Export comprehensive audit report for external auditors
	Includes additional metadata and formatting
	"""
	filters = frappe.parse_json(filters) if isinstance(filters, str) else filters
	
	columns, data = execute(filters)
	
	# Add summary statistics
	summary = {
		"total_activities": len(data),
		"unique_users": len(set(row.get("user") for row in data if row.get("user"))),
		"date_range": f"{filters.get('from_date')} to {filters.get('to_date')}",
		"activity_breakdown": {}
	}
	
	# Count by activity type
	for row in data:
		activity = row.get("activity_type", "Unknown")
		summary["activity_breakdown"][activity] = summary["activity_breakdown"].get(activity, 0) + 1
	
	return {
		"summary": summary,
		"data": data,
		"columns": columns
	}


@frappe.whitelist()
def get_user_activity_stats(user, from_date=None, to_date=None):
	"""Get detailed activity statistics for a specific user"""
	query = """
		SELECT
			cat.activity_type,
			COUNT(*) as count,
			MIN(cat.timestamp) as first_activity,
			MAX(cat.timestamp) as last_activity,
			COUNT(DISTINCT cat.document_name) as unique_documents
		FROM `tabCash Document Audit Trail` cat
		WHERE cat.user = %(user)s
	"""
	
	params = {"user": user}
	
	if from_date:
		query += " AND DATE(cat.timestamp) >= %(from_date)s"
		params["from_date"] = from_date
	
	if to_date:
		query += " AND DATE(cat.timestamp) <= %(to_date)s"
		params["to_date"] = to_date
	
	query += """
		GROUP BY cat.activity_type
		ORDER BY count DESC
	"""
	
	stats = frappe.db.sql(query, params, as_dict=1)
	
	return stats


@frappe.whitelist()
def check_compliance_issues(from_date, to_date, company=None):
	"""
	Run automated compliance checks on audit trail
	Returns list of potential compliance issues
	"""
	issues = []
	
	filters = {
		"from_date": from_date,
		"to_date": to_date
	}
	
	if company:
		filters["company"] = company
	
	columns, data = execute(filters)
	
	# Check 1: Documents modified after verification
	verified_docs = {}
	for row in data:
		if row.get("activity_type") in ["Manual Verification", "ERP Verification", "Final Verification"]:
			doc_key = f"{row.get('document_type')}::{row.get('document_name')}"
			if doc_key not in verified_docs or row.get("timestamp") > verified_docs[doc_key]:
				verified_docs[doc_key] = row.get("timestamp")
	
	for row in data:
		if row.get("activity_type") == "Document Modification":
			doc_key = f"{row.get('document_type')}::{row.get('document_name')}"
			if doc_key in verified_docs and row.get("timestamp") > verified_docs[doc_key]:
				issues.append({
					"type": "Modification After Verification",
					"severity": "High",
					"document": row.get("document_name"),
					"details": f"Document modified at {row.get('timestamp')} after verification at {verified_docs[doc_key]}",
					"user": row.get("user")
				})
	
	# Check 2: Excessive cancellations by single user
	cancellation_count = {}
	for row in data:
		if row.get("activity_type") == "Document Cancellation":
			user = row.get("user", "Unknown")
			cancellation_count[user] = cancellation_count.get(user, 0) + 1
	
	for user, count in cancellation_count.items():
		if count > 10:  # Threshold
			issues.append({
				"type": "Excessive Cancellations",
				"severity": "Medium",
				"user": user,
				"details": f"User has {count} document cancellations in the period",
				"document": None
			})
	
	# Check 3: Balance updates without subsequent verification
	balance_updates = [row for row in data if row.get("activity_type") == "Balance Update"]
	verifications = [row for row in data if "Verification" in row.get("activity_type", "")]
	
	if len(balance_updates) > 0:
		verification_ratio = len(verifications) / len(balance_updates)
		if verification_ratio < 0.5:  # Less than 50% verified
			issues.append({
				"type": "Low Verification Rate",
				"severity": "High",
				"details": f"Only {verification_ratio*100:.1f}% of balance updates are verified ({len(verifications)}/{len(balance_updates)})",
				"user": None,
				"document": None
			})
	
	# Check 4: Activities outside business hours
	for row in data:
		timestamp = get_datetime(row.get("timestamp"))
		hour = timestamp.hour
		
		if hour < 6 or hour > 22:  # Outside 6 AM - 10 PM
			issues.append({
				"type": "Off-Hours Activity",
				"severity": "Low",
				"document": row.get("document_name"),
				"details": f"Activity at {timestamp.strftime('%Y-%m-%d %H:%M:%S')} (hour: {hour})",
				"user": row.get("user")
			})
	
	return issues
